File: src/project-bootstrap.js
import fs from 'fs/promises';
import path from 'path';
import { execFile } from 'child_process';
import { promisify } from 'util';

const execFileAsync = promisify(execFile);

/**
 * Detects the Godot version from the godot binary in PATH
 * @returns {Promise<string|null>} Version string or null if not found
 */
async function detectGodotVersion() {
    try {
        const { stdout } = await execFileAsync('godot', ['--version']);
        // Parse version from output like "4.5.0.stable.official.abcd1234"
        const match = stdout.trim().match(/^(\d+\.\d+)/);
        if (match) {
            return match[1]; // Returns "4.5" from "4.5.0.stable.official.abcd1234"
        }
    } catch (err) {
        // godot binary not found or failed to execute
        return null;
    }
    return null;
}

/**
 * Bootstraps a new Godot project with optional GUT (Godot Unit Testing) support
 * @param {string} projectPath - Absolute path where the project should be created
 * @param {object} options - Configuration options
 * @param {string} options.projectName - Name of the project
 * @param {string} options.godotVersion - Godot version (auto-detected from PATH if not provided)
 * @param {boolean} options.includeGUT - Whether to include GUT testing framework
 * @param {string} options.gutVersion - GUT version (defaults to "9.5.1" for Godot 4.x)
 * @returns {Promise<object>} Result with created files and configuration
 */
export async function bootstrapProject(projectPath, options = {}) {
    // Auto-detect Godot version if not provided
    let detectedVersion = options.godotVersion;
    if (!detectedVersion) {
        detectedVersion = await detectGodotVersion();
        if (!detectedVersion) {
            detectedVersion = "4.5"; // Fallback default
        }
    }

    const {
        projectName = path.basename(projectPath),
        godotVersion = detectedVersion,
        includeGUT = false,
        gutVersion = "9.5.1"
    } = options;

    // Check if directory exists
    try {
        await fs.access(projectPath);
        throw new Error(`Directory already exists: ${projectPath}`);
    } catch (err) {
        if (err.code !== 'ENOENT') throw err;
    }

    // Create project directory
    await fs.mkdir(projectPath, { recursive: true });

    const createdFiles = [];

    try {
        // 1. Create project.godot
        const projectGodotContent = generateProjectGodot(projectName, godotVersion, includeGUT);
        const projectGodotPath = path.join(projectPath, 'project.godot');
        await fs.writeFile(projectGodotPath, projectGodotContent);
        createdFiles.push('project.godot');

        // 2. Create basic icon.svg (Godot default)
        const iconPath = path.join(projectPath, 'icon.svg');
        await fs.writeFile(iconPath, generateDefaultIcon());
        createdFiles.push('icon.svg');

        // 3. Create basic directory structure
        await fs.mkdir(path.join(projectPath, 'scenes'), { recursive: true });
        await fs.mkdir(path.join(projectPath, 'scripts'), { recursive: true });
        createdFiles.push('scenes/');
        createdFiles.push('scripts/');

        // 4. Create .gitignore for Godot
        const gitignorePath = path.join(projectPath, '.gitignore');
        await fs.writeFile(gitignorePath, generateGitignore());
        createdFiles.push('.gitignore');

        // 5. Setup GUT if requested
        if (includeGUT) {
            await setupGUT(projectPath, gutVersion);
            createdFiles.push('test/');
            createdFiles.push('.gutconfig.json');
            createdFiles.push('addons/gut/');
        }

        return {
            success: true,
            projectPath,
            projectName,
            createdFiles,
            includesGUT: includeGUT,
            message: `Successfully bootstrapped Godot project at ${projectPath}`
        };

    } catch (error) {
        // Cleanup on failure
        try {
            await fs.rm(projectPath, { recursive: true, force: true });
        } catch (cleanupErr) {
            console.error('Failed to cleanup after error:', cleanupErr);
        }
        throw error;
    }
}

/**
 * Generates the content for project.godot file
 */
function generateProjectGodot(projectName, godotVersion, includeGUT) {
    let content = `; Engine configuration file.
; It's best edited using the editor UI and not directly,
; since the parameters that go here are not all obvious.
;
; Format:
;   [section] ; section goes between []
;   param=value ; assign values to parameters

config_version=5

[application]

config/name="${projectName}"
config/features=PackedStringArray("${godotVersion}")
config/icon="res://icon.svg"
`;

    if (includeGUT) {
        content += `
[editor_plugins]

enabled=PackedStringArray("res://addons/gut/plugin.cfg")
`;
    }

    return content;
}

/**
 * Generates a basic Godot icon.svg
 */
function generateDefaultIcon() {
    // This is a simplified version of Godot's default icon
    return `<svg height="128" width="128" xmlns="http://www.w3.org/2000/svg">
  <rect fill="#363d52" height="128" rx="6.4" width="128"/>
  <g transform="scale(.101) translate(122 122)">
    <g fill="#fff">
      <path d="m464 428c0 22-18 40-40 40h-336c-22 0-40-18-40-40v-336c0-22 18-40 40-40h336c22 0 40 18 40 40z"/>
      <path d="m672 96v336c0 44-36 80-80 80h-16c0 44-36 80-80 80h-336c-44 0-80-36-80-80v-336c0-44 36-80 80-80h16c0-44 36-80 80-80h336c44 0 80 36 80 80z" fill="none" stroke="#fff" stroke-width="40"/>
    </g>
  </g>
</svg>`;
}

/**
 * Generates .gitignore for Godot projects
 */
function generateGitignore() {
    return `# Godot 4+ specific ignores
.godot/

# Godot-specific ignores
.import/
export.cfg
export_presets.cfg

# Imported translations (automatically generated from CSV files)
*.translation

# Mono-specific ignores
.mono/
data_*/
mono_crash.*.json

# System/tool-specific ignores
.DS_Store
*~
`;
}

/**
 * Sets up GUT (Godot Unit Testing) in the project
 */
async function setupGUT(projectPath, gutVersion) {
    // Create test directory
    const testDir = path.join(projectPath, 'test', 'unit');
    await fs.mkdir(testDir, { recursive: true });

    // Create .gutconfig.json
    const gutConfig = {
        dirs: ["res://test/unit/"],
        double_strategy: "partial",
        ignore_pause: false,
        include_subdirs: true,
        inner_class: "",
        log_level: 1,
        opacity: 100,
        prefix: "test_",
        selected: "",
        should_exit: false,
        should_maximize: false,
        suffix: ".gd",
        tests: [],
        unit_test_name: ""
    };

    await fs.writeFile(
        path.join(projectPath, '.gutconfig.json'),
        JSON.stringify(gutConfig, null, 2)
    );

    // Create a sample test file
    const sampleTestPath = path.join(testDir, 'test_example.gd');
    const sampleTestContent = `extends GutTest

# This is an example test file
# GUT will automatically discover and run tests in files matching the pattern test_*.gd

func test_example_assertion():
    # Basic assertion examples
    assert_true(true, "This should pass")
    assert_eq(2 + 2, 4, "Math should work")

func test_example_failing():
    # This test is pending and won't run
    pending("Implement this test later")
`;
    await fs.writeFile(sampleTestPath, sampleTestContent);

    // Download and extract GUT
    await downloadGUT(projectPath, gutVersion);
}

/**
 * Downloads GUT from GitHub releases
 */
async function downloadGUT(projectPath, version) {
    const addonsDir = path.join(projectPath, 'addons');
    await fs.mkdir(addonsDir, { recursive: true });

    // For now, we'll provide instructions to download manually
    // In a production version, we could use fetch/download libraries
    const readmePath = path.join(addonsDir, 'GUT_INSTALLATION.md');
    const instructions = `# GUT Installation

To complete the GUT setup:

1. Download GUT ${version} from: https://github.com/bitwes/Gut/releases/tag/v${version}
2. Extract the \`addons/gut\` directory from the zip file
3. Copy it to: ${addonsDir}/gut/
4. Restart Godot to enable the plugin

Alternatively, you can install GUT via the AssetLib in Godot Editor.

Once installed, run tests with:
- From Editor: Go to the "GUT" panel at the bottom
- From Command Line: \`godot --path . -s addons/gut/gut_cmdln.gd\`
`;

    await fs.writeFile(readmePath, instructions);

    // Try to download using curl/wget if available
    try {
        const gutZipUrl = `https://github.com/bitwes/Gut/archive/refs/tags/v${version}.zip`;
        const zipPath = path.join(projectPath, 'gut_temp.zip');

        // Try curl first
        try {
            await execFileAsync('curl', ['-L', '-o', zipPath, gutZipUrl]);
        } catch {
            // Fallback to wget
            await execFileAsync('wget', ['-O', zipPath, gutZipUrl]);
        }

        // Try to extract (requires unzip)
        try {
            const extractDir = path.join(projectPath, 'gut_extract');
            await execFileAsync('unzip', ['-q', zipPath, '-d', extractDir]);

            // Find and copy the addons/gut directory
            const gutSourceDir = path.join(extractDir, `Gut-${version}`, 'addons', 'gut');
            const gutDestDir = path.join(addonsDir, 'gut');

            // Use cp -r to copy recursively
            await execFileAsync('cp', ['-r', gutSourceDir, gutDestDir]);

            // Cleanup
            await fs.rm(zipPath, { force: true });
            await fs.rm(extractDir, { recursive: true, force: true });
            await fs.rm(readmePath, { force: true }); // Remove instructions since we succeeded

        } catch (extractErr) {
            console.error('Could not auto-extract GUT. Manual installation required.', extractErr);
        }
    } catch (downloadErr) {
        console.error('Could not auto-download GUT. Manual installation required.', downloadErr);
        // Instructions file already created above
    }
}

File: src/godot_mcp/project_bootstrap.py
"""Bootstrap new Godot projects with optional GUT (Godot Unit Testing) support."""

import io
import json
import re
import shutil
import subprocess
import sys
import zipfile
from pathlib import Path

import httpx


def log(*args):
    print(*args, file=sys.stderr, flush=True)


def detect_godot_version() -> str | None:
    """Detect the Godot version from the godot binary in PATH (e.g. '4.5')."""
    try:
        result = subprocess.run(["godot", "--version"], capture_output=True,
                                text=True, timeout=30)
        match = re.match(r"^(\d+\.\d+)", result.stdout.strip())
        if match:
            return match.group(1)
    except (OSError, subprocess.TimeoutExpired):
        return None
    return None


async def bootstrap_project(project_path: str, project_name: str | None = None,
                            godot_version: str | None = None, include_gut: bool = False,
                            gut_version: str = "9.5.1") -> dict:
    if not godot_version:
        godot_version = detect_godot_version() or "4.5"

    project_dir = Path(project_path)
    project_name = project_name or project_dir.name

    project_dir.mkdir(parents=True, exist_ok=True)

    entries = [p.name for p in project_dir.iterdir()]
    if "project.godot" in entries:
        raise FileExistsError(f"Godot project already exists at: {project_path}")
    if entries:
        log(f"Warning: Directory is not empty. Bootstrapping in existing directory: {project_path}")

    created_files = []
    try:
        (project_dir / "project.godot").write_text(
            generate_project_godot(project_name, godot_version, include_gut))
        created_files.append("project.godot")

        (project_dir / "icon.svg").write_text(generate_default_icon())
        created_files.append("icon.svg")

        (project_dir / "scenes").mkdir(exist_ok=True)
        (project_dir / "scripts").mkdir(exist_ok=True)
        created_files += ["scenes/", "scripts/"]

        (project_dir / ".gitignore").write_text(generate_gitignore())
        created_files.append(".gitignore")

        if include_gut:
            await setup_gut(project_dir, gut_version)
            created_files += ["test/", ".gutconfig.json", "addons/gut/"]

        return {
            "success": True,
            "projectPath": str(project_dir),
            "projectName": project_name,
            "createdFiles": created_files,
            "includesGUT": include_gut,
            "message": f"Successfully bootstrapped Godot project at {project_path}",
        }
    except Exception:
        shutil.rmtree(project_dir, ignore_errors=True)
        raise


def generate_project_godot(project_name: str, godot_version: str, include_gut: bool) -> str:
    content = f"""; Engine configuration file.
; It's best edited using the editor UI and not directly,
; since the parameters that go here are not all obvious.
;
; Format:
;   [section] ; section goes between []
;   param=value ; assign values to parameters

config_version=5

[application]

config/name="{project_name}"
config/features=PackedStringArray("{godot_version}")
config/icon="res://icon.svg"
"""
    if include_gut:
        content += """
[editor_plugins]

enabled=PackedStringArray("res://addons/gut/plugin.cfg")
"""
    return content


def generate_default_icon() -> str:
    # Simplified version of Godot's default icon
    return """<svg height="128" width="128" xmlns="http://www.w3.org/2000/svg">
  <rect fill="#363d52" height="128" rx="6.4" width="128"/>
  <g transform="scale(.101) translate(122 122)">
    <g fill="#fff">
      <path d="m464 428c0 22-18 40-40 40h-336c-22 0-40-18-40-40v-336c0-22 18-40 40-40h336c22 0 40 18 40 40z"/>
      <path d="m672 96v336c0 44-36 80-80 80h-16c0 44-36 80-80 80h-336c-44 0-80-36-80-80v-336c0-44 36-80 80-80h16c0-44 36-80 80-80h336c44 0 80 36 80 80z" fill="none" stroke="#fff" stroke-width="40"/>
    </g>
  </g>
</svg>"""


def generate_gitignore() -> str:
    return """# Godot 4+ specific ignores
.godot/

# Godot-specific ignores
.import/
export.cfg
export_presets.cfg

# Imported translations (automatically generated from CSV files)
*.translation

# Mono-specific ignores
.mono/
data_*/
mono_crash.*.json

# System/tool-specific ignores
.DS_Store
*~
"""


async def setup_gut(project_dir: Path, gut_version: str):
    test_dir = project_dir / "test" / "unit"
    test_dir.mkdir(parents=True, exist_ok=True)

    gut_config = {
        "dirs": ["res://test/unit/"],
        "double_strategy": "partial",
        "ignore_pause": False,
        "include_subdirs": True,
        "inner_class": "",
        "log_level": 1,
        "opacity": 100,
        "prefix": "test_",
        "selected": "",
        "should_exit": False,
        "should_maximize": False,
        "suffix": ".gd",
        "tests": [],
        "unit_test_name": "",
    }
    (project_dir / ".gutconfig.json").write_text(json.dumps(gut_config, indent=2))

    sample_test = """extends GutTest

# This is an example test file
# GUT will automatically discover and run tests in files matching the pattern test_*.gd

func test_example_assertion():
    # Basic assertion examples
    assert_true(true, "This should pass")
    assert_eq(2 + 2, 4, "Math should work")

func test_example_failing():
    # This test is pending and won't run
    pending("Implement this test later")
"""
    (test_dir / "test_example.gd").write_text(sample_test)

    await download_gut(project_dir, gut_version)


async def download_gut(project_dir: Path, version: str):
    addons_dir = project_dir / "addons"
    addons_dir.mkdir(exist_ok=True)

    readme_path = addons_dir / "GUT_INSTALLATION.md"
    readme_path.write_text(f"""# GUT Installation

To complete the GUT setup:

1. Download GUT {version} from: https://github.com/bitwes/Gut/releases/tag/v{version}
2. Extract the `addons/gut` directory from the zip file
3. Copy it to: {addons_dir}/gut/
4. Restart Godot to enable the plugin

Alternatively, you can install GUT via the AssetLib in Godot Editor.

Once installed, run tests with:
- From Editor: Go to the "GUT" panel at the bottom
- From Command Line: `godot --path . -s addons/gut/gut_cmdln.gd`
""")

    try:
        url = f"https://github.com/bitwes/Gut/archive/refs/tags/v{version}.zip"
        async with httpx.AsyncClient(follow_redirects=True, timeout=120) as client:
            response = await client.get(url)
            response.raise_for_status()

        extract_dir = project_dir / "gut_extract"
        with zipfile.ZipFile(io.BytesIO(response.content)) as zf:
            zf.extractall(extract_dir)

        gut_source = extract_dir / f"Gut-{version}" / "addons" / "gut"
        shutil.copytree(gut_source, addons_dir / "gut")

        shutil.rmtree(extract_dir, ignore_errors=True)
        readme_path.unlink(missing_ok=True)  # Instructions unnecessary on success
    except Exception as e:
        log(f"Could not auto-download GUT. Manual installation required. {e}")
        # Instructions file already created above
